tmt/utils/filesystem.py

```diff
@@ -0,0 +1,161 @@
+    # Use a persistent cache for hardlinks
+    cache_dir = Path(tempfile.gettempdir()) / 'tmt_file_cache'
+    cache_dir.mkdir(exist_ok=True)
+
+    if not dst_path.exists():
+        dst_path.mkdir(parents=True, exist_ok=True)
+
+    hardlink_count = 0
+    regular_copy_count = 0
+
+    for item in src_path.rglob('*'):
+        relative_path = item.relative_to(src_path)
+        dst_item = dst_path / relative_path
+
+        # Create a path in cache using a hash of the relative path to avoid path length issues
+        cache_key = relative_path_to_cache_key(relative_path)
+        cache_path = cache_dir / cache_key
+
+        if item.is_dir():
+            dst_item.mkdir(parents=True, exist_ok=True)
+        elif item.is_file():
+            # Create parent directories if they don't exist
+            dst_item.parent.mkdir(parents=True, exist_ok=True)
+
+            # Use cached version if it exists and has identical content
+            content_identical = False
+            if cache_path.exists():
+                try:
+                    content_identical = filecmp.cmp(item, cache_path, shallow=False)
+                except OSError:
+                    # Handle cases where the file might be inaccessible
+                    content_identical = False
+
+            if content_identical:
+                try:
+                    # Remove destination if it exists (needed for hardlink)
+                    if dst_item.exists():
+                        dst_item.unlink()
+
+                    # Create hardlink from cache
+                    os.link(cache_path, dst_item)
+                    hardlink_count += 1
+                    continue
+                except (OSError, FileExistsError):
+                    # Fall through to regular copy if hardlink fails
+                    pass
+
+            # Regular copy if hardlink approach didn't work
+            shutil.copy2(item, dst_item)
+            regular_copy_count += 1
+
+            # Update cache with this file
+            try:
+                if cache_path.exists():
+                    cache_path.unlink()
+                os.link(dst_item, cache_path)
+            except OSError:
+                # Ignore cache errors - they shouldn't affect the main operation
+                pass
+
+        elif item.is_symlink():
+            # Handle symbolic links
+            if dst_item.exists() or dst_item.is_symlink():
+                dst_item.unlink()
+            link_target = os.readlink(item)
+            os.symlink(link_target, dst_item)
+
+    if logger:
+        if hardlink_count > 0:
+            logger.debug(
+                f"Directory copied using hardlinks: {hardlink_count} hardlinked, "
+                f"{regular_copy_count} regular copies"
+            )
+        else:
+            logger.debug("Directory copied using regular copy method")
+
+    return
+
+
+def relative_path_to_cache_key(relative_path: Union[str, Path]) -> str:
```

Comment:
`hashlib` is full of hashing functions, why not using one from this standard library package?